### PR TITLE
[tests/modelfile] Enable modelfile unittest @open sesame 06/30 17:00

### PR DESCRIPTION
**Pull request by kparichay**
Enable model file unit tests with newly update LayerV2 to test
building and initialization of the models.

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Review by jijoongmoon — APPROVED** [review]
LGTM. I'm sure you are. are you going to uncomment the test cases later PR?

**Comment by kparichay** [comment]
> LGTM. I'm sure you are. are you going to uncomment the test cases later PR?

Yes, with enabling of each layer, their corresponding unittests will be enabled.

File: nntrainer/graph/network_graph.cpp [hunk 1]
```diff
@@ -324,27 +324,15 @@ void NetworkGraph::setOutputLayers() {
...
-      if (layer_idx->getNumOutputs() == 0) {
-        /** No output layer inplies its the last layer */
-        layer_idx->setOutputLayers({"__exit__"});
-        last_layer_count += 1;
-      } else {
-        /** error for any other layer */
-        throw std::logic_error("Graph node has fewer edges than expected.");
-      }
+    if (layer_idx->getNumOutputs() == 0) {
+      last_layer_count += 1;
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:328]: Q. now, we don't have `__exit__` anymore?

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:328]: yes, __exit__ and __data__ is not needed anymore.

File: nntrainer/graph/network_graph.cpp [hunk 2]
```diff
@@ -793,9 +787,12 @@ int NetworkGraph::initialize(std::shared_ptr<Manager> manager) {
     lptr->setOutputBuffers(in_out);
 #endif
 
+    if (idx == graph.size() - 1)
```
Comment (zhoonit) [inline nntrainer/graph/network_graph.cpp:790]: Wonder what this is for, we are skipping the last layer?

Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:790]: Yes, for last layer, there is no need to update the input_map anymore. So, it is skipped.
added a comment here.

File: test/unittest/unittest_nntrainer_modelfile.cpp [hunk 3]
```diff
@@ -648,10 +655,10 @@ TEST(nntrainerIniTest, backbone_p_20) {
  * it should be referred relative to the .ini
  *
  */
-TEST(nntrainerIniTest, backbone_relative_to_ini_p) {
-  ScopedIni b{getResPath("base"), {nw_base, batch_normal}};
+TEST(nntrainerIniTest, DISABLED_backbone_relative_to_ini_p) {
+  ScopedIni b{getResPath("base"), {nw_base_cross, batch_normal}};
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_modelfile.cpp:659]: Wonder why this is being changed.

File: test/unittest/unittest_nntrainer_modelfile.cpp [hunk 4]
```diff
@@ -648,10 +655,10 @@ TEST(nntrainerIniTest, backbone_p_20) {
...
   ScopedIni s{getResPath("original"),
-              {nw_base + "loss=mse", adam, input,
+              {nw_base_cross + "loss=mse", adam, input,
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_modelfile.cpp:661]: I think this will override nw_base to mse.

File: test/unittest/unittest_nntrainer_modelfile.cpp [hunk 5]
```diff
@@ -165,8 +165,11 @@ TEST_P(nntrainerIniTest, initThreetime_n) {
 
 /// @todo add run test could be added with iniTest flag to control skip
 static nntrainer::IniSection nw_base("model", "Type = NeuralNetwork | "
-                                              "batch_size = 32 | "
-                                              "loss = cross");
+                                              "batch_size = 32");
+
+static nntrainer::IniSection nw_base_cross("model", "Type = NeuralNetwork | "
+                                                    "batch_size = 32 | "
+                                                    "loss = cross");
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_modelfile.cpp:172]: Could you add tests that loss is at the last layer explicitly?

File: test/unittest/unittest_nntrainer_modelfile.cpp (hunk 3 above)
Comment (kparichay) [reply test/unittest/unittest_nntrainer_modelfile.cpp:659]: backbone unittests are disabled till backbone layers are implemented.
For nw_base, as nw_base is redefined to be without any optimizer, nw_base usage is replaced with nw_base_cross (to maintain original functionality especially for unittests which cannot be tested now).

File: test/unittest/unittest_nntrainer_modelfile.cpp (hunk 4 above)
Comment (kparichay) [reply test/unittest/unittest_nntrainer_modelfile.cpp:661]: this maintains the original code implemented (nw_base(old) == nw_base_cross)
And yes, hopefully this will override to mse.

File: test/unittest/unittest_nntrainer_modelfile.cpp (hunk 5 above)
Comment (kparichay) [reply test/unittest/unittest_nntrainer_modelfile.cpp:172]: I will add it as a separate PR soon? The purpose of this patch is to enable existing tests and check its working with implemented code.

**Review by zhoonit — APPROVED** [review]
LGTM

File: test/unittest/unittest_nntrainer_modelfile.cpp (hunk 5 above)
Comment (zhoonit) [reply test/unittest/unittest_nntrainer_modelfile.cpp:172]: Alright

Comment (kparichay) [reply test/unittest/unittest_nntrainer_modelfile.cpp:172]: added #1363. Please check.
